RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,644 @@
+query with `maybe_get_output()` to register any outputs necessary.
+
+## Piggybacking other improvements
+
+Ports to structured kernels must be done individually by hand.  Because
+the porting process is already labor intensive, we should do other
+improvements "while the patient is open".  Here are some candidate
+improvements which should consider applying
+
+### Removing mutable references from out arguments
+
+Historically, out and inplace variants of functions take a `Tensor&`
+rather than a `const Tensor&`.  This convention has lead to no end
+of confusion for kernel writers, who incorrectly surmise that given
+a mutable reference `Tensor& out`, one may assign the output by
+writing `out = ... some expression ...`  (this doesn't work).
+
+The absence of the const modifier is currently relied upon by
+template metaprogramming machinery (to detect if arguments are out
+tensors or not); however, because the implementations of structured
+kernels are a layer below the operator registration layer, the
+const modifier can be eliminated from the `TORCH_IMPL_FUNC` API
+without requiring the rest of the system to be updated.
+
+One implication of this change is that the out parameter cannot be
+easily passed to existing public API that requires a mutable reference.
+This can be easily remedied by updating the existing APIs to accept
+const references and not only mutable references.
+
+### Type refinement
+
+Once we have dispatched to a CPU kernel, we know that the tensor in
+question is in fact a CPU tensor, and not (for example) a CUDA tensor.
+However, this information is not retained inside of the body of the
+kernel, and so if a user makes a method call or regular `at::` namespace
+function call, the dispatcher still must inspect the type tag to
+rediscover, yes, indeed, we still have a CPU tensor.
+
+One promising approach to solving this problem is to refine the type of
+a tensor from `Tensor` to `CPUTensor`, where a `CPUTensor` represents a
+tensor that is statically known to be a CPU Tensor.  Operations
+(functions and methods) on `CPUTensor` bypass dispatching and go
+directly to the CPU implementations in question.  `const CPUTensor&` can
+be defined to implicitly convert into `const Tensor&`, which means
+existing APIs that don't know how to short circuit can continue to do
+pre-existing behavior.
+
+The primary consequence of making this change immediately is we must
+immediately create a CPUTensor class with enough methods to cover the
+usual surface area (even if those methods don't apply any performance
+optimization).  With code generation this should not be too much code.
+This would also require the creation of a CPUTensorRef class to ensure
+that CPUTensors can be created from `const Tensor&` without incurring
+a reference count bump).
+
+One question is wheter or not the existence of CPUTensor means we should
+eliminate the `at::cpu::` namespace (as they serve near equivalent purposes;
```

Comment:
`at::cpu` seems much less risky around the edges than `CPUTensor`[1] - if they serve near equivalent purposes, what's an example that motivates adding CPUTensor either instead of or in addition to `at::cpu`?

[1] stemming from CPUTensor not reeeally being a subtype of Tensor. A stray call to something like `set_storage_and_dtype` would damage a CPUTensor that's been upcast to a Tensor (or provoke a runtime error, if we've put safety measures in place).

Reply:
It is an ergonomics thing (but important ergonomics): with `at::cpu`, you have to remember to use the CPU only function (and you have to make sure not to accidentally call it with a non-CPU tensor). CPUTensor ensures that the CPU tensor is used, and you only need to prove you have CPU tensor when you initially construct the CPUTensor.